src/legacy/ui/public/saved_objects/saved_object.js

```diff
@@ -326,7 +326,7 @@ export function SavedObjectProvider(Promise, Private, confirmModalPromise, index
           references.push({
             name: refName,
             type: 'index-pattern',
-            id: indexId,
+            id: indexId || searchSourceFields.index,
```

Comment:
is it a necessary change? we ready anyway we read `id` later https://github.com/elastic/kibana/pull/42974/files#diff-e8fcd698a977c19072c4c82d0d4eb40bR149
probably we should throw if `id` is not present?

Reply:
According to the source code comments when `hydrateIndexPattern()` is called index can be either a string or an IndexPattern, but after the call it's converted to an IndexPattern. So I think this can happen if a Saved Object is created and then `_serialize()` is called before `hydrateIndexPattern` is called. This might happen here https://github.com/elastic/kibana/pull/42974/files#diff-28408d72ee9bcd7cf7c986c55a9212d7R279

But I'll take a deeper look and also refactor this a bit (if it's necessary). 

```
const indexId = typeof(searchSourceFields.index) == 'string' : searchSourceFields.index : searchSourceFields.index.id;
```

Reply:
My comment above wasn't 100% accurate...

group by ID https://github.com/elastic/kibana/pull/42974/files#diff-28408d72ee9bcd7cf7c986c55a9212d7L276 used to use `obj.searchSource.getOwnField('index')` (a string) which would only group by searchSource indexes and not other referenced indexes. To get access to the other referenced indexes I need to call `_serialize()`. When `serialize()` is called `searchSource.getFields('index')` is a string because these index patterns don't exist so they never get resolved to an IndexPattern object in `hydrateIndexPattern()`.

I think this behaviour change makes sense, when a saved object is serialized (which normally happens before saving) we should persist the id of referenced indexes even if these don't exist in Elasticsearch.